### ISSUE TITLE
Possibility to disable linter for version.js

### ISSUE THREAD

**Issue by nekhebit** [axelpale/genversion#19]
Thank you for your work on `genversion`.

I am having an issue with `ESLint` and `genversion`. `ESLint` is changing version.js and it seems that is no option to deactivate linting in `genversion`. 
It would be great to have an option to turn linting off, so `genversion` would create a version.js similar to the following:

```javascript
/* eslint-disable */
/* tslint:disable */
// Generated by genversion.
'use strict';

export const version = '1.4.4';
```
If this is already possible, could someone show me how to activate it? 

Thank you!

**Comment by axelpale** [comment]
Hi and thanks for the issue. In order to advice you properly, could you show how your ESLint is changing version.js? Does it add something to the file? Does your linting rules require something that genversion options cannot provide?

Disabling the linter for the version file in the way of your example could be a good feature for genversion. Endless combinations of linter rules have been troublesome for the version file generation. If genversion inserted eslint-disable line, it would solve most ESLint-related problems. Therefore thanks for the suggestion! Still, I am interested of how your ESLint is changing the version.js.

**Comment by nekhebit** [comment]
Hey, thanks for the speedy answer!

I'm using an ESLint plugin to add license-headers to *.js/*.ts files. 
The plugin checks every file for the license and and adds it if it is missing, as long as it does not have `/*eslint:disable*/` in it.

**Comment by axelpale** [comment]
I see. Thanks for the clarification! The automatic license header does make sense. I remember it is recommended by open source licensing organisations.

Let me think for a solution to this a bit. Probably the best is to add the magic disable line as you suggested. We will see.

**Comment by axelpale** [comment]
Sketching out possible solutions. Any opinions?

## Solution 1: custom header string
```
$ genversion --header "/* eslint-disable*/\n/* tslint:disable */\n// Generated by genversion." version.js
```

Pro: solves many unforeseen situations.
Con: escaping characters, especially the newlines, can become pain on the command line. Not a problem in JS API.

## Solution 2: custom header file
```
$ genversion --header=version_header.js version.js
```

Pro: simple. Use header as written in the header file.
Con: The header file itself becomes a subject of linting rules.

## Solution 3: disable linter
```
$ genversion --disable-linter=eslint version.js
```
or
```
$ genversion --nolint=eslint version.js
```

Pro: clean and simple.
Con: only solves the issue for known linters.

**Comment by nekhebit** [comment]
Thank you for involving me in it! Here are my 2 cents: 

**Solution 1**:  :-1: 
Users can get messy/run into more problems and blame the software for that. Unfortunately this happens more often than one would imagine. :-1: 

-**Solution 3**-:  :-1: 
Nice and clean but, like you said, very restrictive and requires more maintenance than the others. 
In a perfect world, this would be the best solution.  There are so many linters out there and it would be a pity that some people could not use `genversion` because their linter is not featured in it.  

**Solution 2**:  :+1: 
Even though I like the cleanness of Solution 3, I think that this is the one to go. 
Sure, it adds another file, that will be checked and ignored by the linter, but it is  still better than to provide a string like in **_Solution 1_** and, unlike **Solution 3**, allows for the usage of any linter.  Another plus is that it is not as maintenance-intensive as the other 2. 

Another possibility would be to create a hybrid from Solution 2 and 3, in which the user could _either_ disable some linters directly _or_ provide a header file. Although I must admit that this solution feels for me like bad design, as it would only increase the complexity of without adding anything new to the software.  

**Conclusion**
Like you pointed out, all 3 solutions have their merits and flaws. Maybe I oversaw something but, after weighting all pros and contras, I do believe that **Solution 2**  it the one that hits the sweet spot between complexity, maintenance and user freedom.

**Comment by axelpale** [comment]
Nice, thank you for such an elaborate analysis, helps a lot!

We will definitely drop Solution 1. Also, as you expertly pointed out, the probability of maintenance is high for Solution 3. Although I like working on genversion, it is always good to try to avoid unnecessary burden.

Therefore I would iterate on Solution 2. Although the custom header solves the issue at hand, it could solve more. Mere header still requires syntax rules for rest of the file to be handled by the other flags. For example `--semi` for the use of semicolon. Instead, if the user specified a template for the whole version file, it could solve the any remaining syntax issues too. Also, something along these lines did come up as a possibility during issue #17. 

## Solution 2B: custom template file

```
$ genversion --template=version.ejs version.js
```

Where the custom template `version.ejs` could contain something like:

```
/* eslint-disable */
/* tslint:disable */
'use strict';

export const version = '<%= version %>';
```

There are many templating languages and tools available. I believe [EJS](https://ejs.co/) to be one of the most common. It is even implemented in the [Lodash library](https://lodash.com/docs/4.17.15#template). In future, if another templating language is required or requested, the file extension of the template file can be used to automatically detect which tool to apply.

With proper documentation this solution might be easy enough to use. When compared to Solution 2, It also removes possible ambiguity of what would happen after the header lines. As an additional benefit, the template file would have non-js file extension and therefore it would not become another source of linting issues because linters will skip it.

Does Solution 2B solve the issue in your case? Does this solution bear any shortcomings?

**Comment by nekhebit** [comment]
Hey, sorry for the late response! 

I think solution 2B could solve my problem.

Only side effect is that it seems that I would have to add `*.ejs` files to the ignore list of ESlint; that should not be an issue, but it is definitely something that you would have to highlight to users in the documentation. 

I think file type exclusion is a standard linter feature, so the usage of EJS should not cause any problems to the users.
 
https://stackoverflow.com/questions/47575167/disabling-eslint-for-ejs-files

**Comment by axelpale** [comment]
This issue will be solved in genversion v3.2.0 with the use of the new `--template <path>` parameter. The parameter allows users to define a template file that can have any content and structure – as long as it is a valid template.

For the original issue, the following EJS template is likely sufficient:

```
/* eslint-disable */
/* tslint:disable */
'use strict';

export const version = '<%= package.version %>';
```

To apply the template, call genversion like this:

```
$ genversion --template=version.ejs lib/version.js
```

We are planning to support other template engines in addition to EJS but currently it is the only one.
